packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+                // chunks for the same module, relying on
+                // naive hash to have additonal
+                // chunks in case if there are same modules being imported in differnt
+                // origins.
+                let chunk =
+                    imported_module.as_chunk(Vc::upcast(chunking_context), availability_info);
+                let chunk_group = chunking_context.chunk_group(chunk);
+                chunks_hash.insert(imported_raw_str.to_string(), chunk_group);
+                chunk_group
+            };
+
+            dynamic_import_chunks
+                .entry(origin_module)
+                .or_insert_with(Vec::new)
+                .push((imported_raw_str.clone(), chunk));
+        }
+    }
+
+    Ok(Vc::cell(dynamic_import_chunks))
+}
+
+pub(crate) async fn collect_evaluated_chunk_group(
+    chunking_context: Vc<Box<dyn EcmascriptChunkingContext>>,
+    dynamic_import_entries: IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules>,
+    availability_info: Value<AvailabilityInfo>,
+    evaluatable_assets: Vc<EvaluatableAssets>,
+) -> Result<Vc<DynamicImportedChunks>> {
+    let mut chunks_hash: HashMap<String, Vc<OutputAssets>> = HashMap::new();
+    let mut dynamic_import_chunks = IndexMap::new();
+
+    // Iterate over the collected import mappings, and create a chunk for each
+    // dynamic import.
+    for (origin_module, dynamic_imports) in dynamic_import_entries {
+        for (imported_raw_str, imported_module) in dynamic_imports {
+            let chunk = if let Some(chunk) = chunks_hash.get(&imported_raw_str) {
+                *chunk
+            } else {
+                let Some(imported_module) =
+                    Vc::try_resolve_sidecast::<Box<dyn EcmascriptChunkPlaceable>>(imported_module)
+                        .await?
+                else {
+                    bail!("module must be evaluatable");
+                };
+
+                // [Note]: this seems to create a duplicated chunk for the same module to the original import() call
+                // and the explicit chunk we ask in here. So there'll be at least 2
+                // chunks for the same module, relying on
+                // naive hash to have additonal
+                // chunks in case if there are same modules being imported in differnt
+                // origins.
+                let chunk =
+                    imported_module.as_chunk(Vc::upcast(chunking_context), availability_info);
+                let chunk_group = chunking_context.evaluated_chunk_group(chunk, evaluatable_assets);
+                chunks_hash.insert(imported_raw_str.to_string(), chunk_group);
+                chunk_group
+            };
+
+            dynamic_import_chunks
+                .entry(origin_module)
+                .or_insert_with(Vec::new)
+                .push((imported_raw_str.clone(), chunk));
+        }
+    }
+
+    Ok(Vc::cell(dynamic_import_chunks))
+}
```

Comment:
Do we need both implementation here?

Reply:
@jridgewell actually it is my q: the reason having separate is how I did to create chunks via `evaluated_chunk_group` vs `chunk_group`, one for the buildchunkingcontext and one for the EcmascriptChunkingContext, as buildchunkingcontext seems not implementing evaluated chunk group. Does new refactoring consolidated those interfaces? if not, what's the recommended way to reduce duplication?

Reply:
No, we don't change evaluated/non-evaluated. I'm not sure really sure why we're using two in the app code.

It seems most of this duplication is just the normalization using `dynamic_import_chunks`. Could we reduce that?

Reply:
yeah I'm also annoyed most of looping is same, but the actual inner part is different. Tried few but involving some jazz with closures to wrap inner to consolidate loop which was not very looking good.

Reply:
Anyway, updated per suggestion.